Implement RealtimeAgent for Real-Time Conversational AI Support in ag2 Framework (#196)
ag2ai/ag2: website/blog/2024-12-18-RealtimeAgent/index.mdx

---
title: Introducing RealtimeAgent Capabilities in AG2
authors:
  - marklysze
  - sternakt
  - davorrunje
  - davorinrusevljan
tags: [Realtime API, Voice Agents, Swarm Teams, Twilio, AI Tools]

---

![Realtime Agent Swarm](img/realtime_agent_swarm.png)

**TL;DR:**
- **RealtimeAgent** is coming in the next AG2 release, enabling real-time conversational AI.
- Features include real-time voice interactions, seamless task delegation to Swarm teams, and Twilio-based telephony integration.

> - Learn how to integrate twillio and RealtimeAgent into your swarm in this blogpost

Reviewer: ```suggestion
- Learn how to integrate twillio and RealtimeAgent into your swarm in this blogpost.
```

### **Realtime API Support: What's New?**

> We're thrilled to announce the release of **RealtimeAgent**, extending AG2's capabilities to support **real-time conversational AI tasks**. This new feature makes it possible for developers to build agents capable of handling voice-based interactions with minimal latency, integrating OpenAI’s Realtime API, Twilio for telephony, and AutoGen’s Swarm task orchestration.

Reviewer: ```suggestion
We're thrilled to announce the release of **RealtimeAgent**, extending AG2's capabilities to support **real-time conversational AI tasks**. This new experimental feature makes it possible for developers to build agents capable of handling voice-based interactions with minimal latency, integrating OpenAI’s Realtime API, Twilio for telephony, and AutoGen’s Swarm task orchestration.
```

- Features include real-time voice interactions, seamless task delegation to Swarm teams, and Twilio-based telephony integration.
- Learn how to integrate twillio and RealtimeAgent into your swarm in this blogpost

### **Realtime API Support: What's New?**

We're thrilled to announce the release of **RealtimeAgent**, extending AG2's capabilities to support **real-time conversational AI tasks**. This new feature makes it possible for developers to build agents capable of handling voice-based interactions with minimal latency, integrating OpenAI’s Realtime API, Twilio for telephony, and AutoGen’s Swarm task orchestration.

### **Why Realtime API Support Matters**

Traditionally, conversational AI tasks have focused on asynchronous interactions, such as text-based chats. However, the demand for **real-time voice agents** has surged in domains like customer support, healthcare, and virtual assistance. With this update, AG2 takes a leap forward by enabling agents to:

1. **Support Real-Time Voice Interactions**
   Engage in real-time conversations with users through seamless integration with Twilio telephony.

2. **Leverage Swarm Teams for Task Delegation**

>    Delegate complex tasks to AutoGen Swarm teams during a voice interaction, ensuring efficient task management.

Reviewer: ```suggestion
   Delegate complex tasks to AG2 Swarm teams during a voice interaction, ensuring efficient task management.
```

3. **Provide Developer-Friendly Integration**
   Tutorial and streamlined API to make setting up real-time agents more straightforward for developers.

> ### **Key Features of RealtimeAgent and RealtimeOpenAIClient**

Reviewer: 2. RealtimeOpenAIClient is missing.

Reviewer 2: That's an internal class

The key functionalities of this implementation include:

1. **Setting Up a REST Endpoint**
   We define a REST API endpoint, `/incoming-call`, to handle incoming voice calls from Twilio. This endpoint returns a Twilio Markup Language (TwiML) response, connecting the call to Twilio’s Media Stream for real-time audio data transfer.

```python
app.api_route("/incoming-call", methods=["GET", "POST"])
async def handle_incoming_call(request: Request):
    """Handle incoming call and return TwiML response to connect to Media Stream."""
    response = VoiceResponse()
    response.say("Please wait while we connect your call to the AI voice assistant.")
    response.pause(length=1)
    response.say("O.K. you can start talking!")
    host = request.url.hostname
    connect = Connect()
    connect.stream(url=f"wss://{host}/media-stream")
    response.append(connect)
    return HTMLResponse(content=str(response), media_type="application/xml")
```
2. **WebSocket Media Stream**

>    A WebSocket endpoint, `/media-stream`, is established to manage real-time audio communication between Twilio and OpenAI's API. This allows audio data to flow seamlessly, enabling the RealtimeAgent to process and respond to user queries.

Reviewer: ```suggestion
   A WebSocket endpoint, `/media-stream`, is established to manage real-time audio communication between Twilio and a realtime model inference client such as OpenAI's realtime API. This allows audio data to flow seamlessly, enabling the RealtimeAgent to process and respond to user queries.
```

```python
@app.websocket("/media-stream")
async def handle_media_stream(websocket: WebSocket):

>     """Handle WebSocket connections between Twilio and OpenAI."""

Reviewer: ```suggestion
    """Handle WebSocket connections between Twilio and realtime model inference client."""
```

    await websocket.accept()
    ...
```
3. **Initializing the RealtimeAgent**
   Inside the WebSocket handler, we instantiate the **RealtimeAgent** with the following components:
   - **Name**: The identifier for the agent (e.g., `Customer_service_Bot`).

>    - **LLM Configuration**: The configuration for the underlying OpenAI model that powers the agent.

Reviewer: ```suggestion
   - **LLM Configuration**: The configuration for the underlying realtime model inference client that powers the agent.
```

   )
```
5. **Running the RealtimeAgent**
   The `run()` method is invoked to start the RealtimeAgent, enabling it to handle real-time voice interactions and delegate tasks to the registered Swarm agents.
```python
   await realtime_agent.run()
```
Here is the full code for this integration:

```python
app = FastAPI(title="Realtime Swarm Agent Chat", version="0.1.0")

@app.get("/", response_class=JSONResponse)
async def index_page():
    return {"message": "Twilio Media Stream Server is running!"}

@app.api_route("/incoming-call", methods=["GET", "POST"])
async def handle_incoming_call(request: Request):
    """Handle incoming call and return TwiML response to connect to Media Stream."""
    response = VoiceResponse()
    response.say("Please wait while we connect your call to the AI voice assistant.")
    response.pause(length=1)
    response.say("O.K. you can start talking!")
    host = request.url.hostname
    connect = Connect()
    connect.stream(url=f"wss://{host}/media-stream")
    response.append(connect)
    return HTMLResponse(content=str(response), media_type="application/xml")

@app.websocket("/media-stream")
async def handle_media_stream(websocket: WebSocket):

>     """Handle WebSocket connections between Twilio and OpenAI."""

Reviewer: ```suggestion
    """Handle WebSocket connections between Twilio and realtime model inference client."""
```
